front/components/vaults/VaultCreateAppModal.tsx

```diff
@@ -0,0 +1,178 @@
+import type { APIError, AppVisibility, WorkspaceType } from "@dust-tt/types";
+import { useContext, useState } from "react";
+
+import { SendNotificationsContext } from "@app/components/sparkle/Notification";
+import { useApps } from "@app/lib/swr";
+import { MODELS_STRING_MAX_LENGTH } from "@app/lib/utils";
+
+export const VaultCreateAppModal = ({
+  owner,
+  isOpen,
+  setIsOpen,
+}: {
+  owner: WorkspaceType;
+  isOpen: boolean;
+  setIsOpen: (isOpen: boolean) => void;
+}) => {
+  const sendNotification = useContext(SendNotificationsContext);
+
+  const [name, setName] = useState<string | null>(null);
+  const [nameError, setNameError] = useState<string | null>(null);
+
+  const [description, setDescription] = useState<string | null>(null);
+  const [descriptionError, setDescriptionError] = useState<string | null>(null);
+
+  const [visibility, setVisibility] = useState<AppVisibility>("private");
+
+  const { apps, mutateApps } = useApps(owner);
+
+  const onSave = async () => {
+    if (!name || name.trim() === "") {
+      setNameError("Name is required.");
+    } else if (!name.match(/^[a-zA-Z0-9._-]+$/)) {
+      setNameError(
+        "App name must only contain letters, numbers, and the characters `._-`"
+      );
+    } else if (apps.find((app) => app.name === name)) {
+      setNameError("An App with this name already exists.");
+    }
+
+    if (!description || description.trim() === "") {
+      setDescriptionError(
+        "A description is required for your app to be selectable in the Assistant Builder."
+      );
+    }
+
+    if (name && description && !nameError && !descriptionError) {
+      const res = await fetch(`/api/w/${owner.sId}/apps`, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          name: name.slice(0, MODELS_STRING_MAX_LENGTH),
+          description: description.slice(0, MODELS_STRING_MAX_LENGTH),
+          visibility,
+        }),
+      });
+      if (res.ok) {
+        await mutateApps();
+        setIsOpen(false);
+        // @todo Daph next PR: rework the modal detail so we can propertly redirect
+        // on the edition view.
+        // const response: PostAppResponseBody = await res.json();
+        // const { app } = response;
+        // await router.push(`/w/${owner.sId}/a/${app.sId}`);
+
+        sendNotification({
+          type: "success",
+          title: "Successfully created app",
+          description: "App was successfully created.",
+        });
+      } else {
+        const err: { error: APIError } = await res.json();
+        sendNotification({
+          title: "Error Saving App",
+          type: "error",
+          description: `Error: ${err.error.message}`,
+        });
+      }
+    }
+  };
+
+  return (
+    <Modal
+      isOpen={isOpen}
+      onClose={() => {
+        setIsOpen(false);
+      }}
+      onSave={async () => {
+        await onSave();
+      }}
```

Comment:
Let's avoid anonymous functions when they are not needed, it will create a new instance of that function every time the parent component re-renders. Which will cause unnecessary re-rendering of child components because React sees the function as a new prop due to its reference changing.

Reply:
Fixed in another PR